identity-apps-core/apps/x509-certificate-authentication-portal/src/main/webapp/includes/branding-preferences.jsp

```diff
@@ -0,0 +1,669 @@
+<%--
+  ~ Copyright (c) 2024, WSO2 LLC. (https://www.wso2.com).
+  ~
+  ~ WSO2 LLC. licenses this file to you under the Apache License,
+  ~ Version 2.0 (the "License"); you may not use this file except
+  ~ in compliance with the License.
+  ~ You may obtain a copy of the License at
+  ~
+  ~    http://www.apache.org/licenses/LICENSE-2.0
+  ~
+  ~ Unless required by applicable law or agreed to in writing,
+  ~ software distributed under the License is distributed on an
+  ~ "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+  ~ KIND, either express or implied.  See the License for the
+  ~ specific language governing permissions and limitations
+  ~ under the License.
+--%>
+
+<%-- Tenant resolve JSP must be included in the calling script inorder to resolve tenant context --%>
+
+<%@ page import="java.io.File" %>
+<%@ page import="java.util.Map" %>
+<%@ page import="java.util.HashMap" %>
+<%@ page import="org.json.JSONObject" %>
+<%@ page import="org.apache.commons.lang.StringUtils" %>
+<%@ page import="org.wso2.carbon.identity.mgt.endpoint.util.client.BrandingPreferenceRetrievalClient" %>
+<%@ page import="org.wso2.carbon.identity.mgt.endpoint.util.client.BrandingPreferenceRetrievalClientException" %>
+<%@ page import="org.wso2.carbon.identity.mgt.endpoint.util.IdentityManagementEndpointConstants" %>
+<%@ page import="javax.servlet.http.HttpServletRequest" %>
+<%@ page import="java.util.*" %>
+
+<%-- TODO: UNIFICATION TASK: This block should be moved to a `locale-code-resolver.jsp` file. And used in `localize` as well. --%>
+<%!
+    /**
+    * Get the user's preferred locale based on the request, cookies, and URL parameters.
+    *
+    * This method determines the user's locale based on the following priority order:
+    * 1. Locale set in a cookie (if available).
+    * 2. Locales specified in the "ui_locales" URL parameter.
+    * 3. Browser's default locale.
+    *
+    * If a valid locale cannot be determined from the cookie or URL parameters, the browser's
+    * default locale is used as the fallback.
+    *
+    * @param request The HTTP servlet request.
+    * @return The user's preferred locale.
+    */
+    public Locale getUserLocale(HttpServletRequest request) {
+        String lang = "en_US"; // Default lang is en_US.
+        String COOKIE_NAME = "ui_lang";
+        String BUNDLE = "org.wso2.carbon.identity.application.authentication.endpoint.i18n.Resources";
+        Locale browserLocale = request.getLocale();
+        Locale userLocale = browserLocale;
+        String uiLocaleFromURL = request.getParameter("ui_locales");
```

Comment:
Have we considered sanitizing request parameters?

Reply:
I believe sanitization is unnecessary in this context. We only use these values to retrieve the resource bundle, and we are not interacting with the database using these values.

Reply:
This code was not added by this pull request; it existed previously. Additionally, there are other locations, such as localize.jsp, where these parameters are used without encoding. If there are any vulnerabilities here, we need to identify all the affected places and address them separately as a security improvement.